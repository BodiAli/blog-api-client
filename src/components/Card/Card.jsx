import { Link, useNavigate, useOutletContext } from "react-router";
import { format } from "date-fns";
import parse from "html-react-parser";
import { toast } from "react-toastify";
import anonymousImage from "../../assets/images/anonymous.jpg";
import noImage from "../../assets/images/no-image.svg";
import styles from "./Card.module.css";

export default function Card({ post, dispatch }) {
  const { token, user } = useOutletContext();
  const navigate = useNavigate();

  async function handleLikePost(e) {
    e.stopPropagation();

    try {
      const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/posts/${post.id}/like`, {
        method: "PATCH",
        headers: {
          Authorization: token,
        },
      });

      if (!res.ok) {
        if (post.liked) {
          throw new Error("Failed to remove like please try again later");
        } else {
          throw new Error("Failed to like post please try again later");
        }
      }

      if (post.liked) {
        toast.success("Like removed successfully!");
      } else {
        toast.success("Post liked successfully!");
      }

      dispatch({
        type: "update-like",
        post: { ...post, liked: !post.liked, likes: post.liked ? post.likes - 1 : post.likes + 1 },
      });
    } catch (error) {
      toast.error(error.message);
    }
  }

  return (
    <div
      onClick={() => {
        navigate(`posts/${post.id}`, { viewTransition: true });
      }}
      className={styles.card}
    >
      <div className={styles.text}>
        <div className={styles.textContainer1}>
          <img
            className={styles.profilePic}
            src={post.User.Profile.profileImgUrl ? post.User.Profile.profileImgUrl : anonymousImage}
            alt={`${post.User.firstName} ${post.User.lastName}'s profile picture`}
          />
          <p className={styles.name}>
            {post.User.firstName} {post.User.lastName}
          </p>
          <p className={styles.createdAt}>{format(post.createdAt, "MMM d, y")}</p>
        </div>
        <div className={styles.textContainer2}>
          <p className={styles.title}>{post.title}</p>
        </div>
        <div className={styles.textContainer3}>{parse(post.content)}</div>
        <div className={styles.textContainer4}>
          {post.Topics.length === 0 ? (
            <span>This post has no topics</span>
          ) : (
            post?.Topics.map((topic, i) => {
              return <p key={i}>{topic.name}</p>;
            })
          )}
        </div>
        <div className={styles.textContainer5}>
          <p data-testid="likes-count">
            <strong>Likes: </strong>
            {post.likes}
          </p>
          {!user ? (
            <p className={styles.loginMessage} data-testid="login-notice">
              Please{" "}
              <Link onClick={(e) => e.stopPropagation()} to="/log-in" viewTransition>
                login
              </Link>{" "}
              to be able to like posts
            </p>
          ) : (
            <button
              data-testid="like-button"
              onClick={handleLikePost}
              className={`${styles.likeButton} ${post.liked ? styles.liked : ""}`}
            ></button>
          )}
        </div>
      </div>
      <div className={styles.imageContainer}>
        <img src={post.imgUrl ? post.imgUrl : noImage} alt="Post cover" />
      </div>
    </div>
  );
}
