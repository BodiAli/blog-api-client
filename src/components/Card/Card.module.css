.card {
  display: grid;
  grid-template-columns: 2fr 1fr;
  align-items: center;
  padding: 1rem;
  gap: 10px;
  transition: background-color 200ms;
  border-radius: 5px;
  &:hover {
    cursor: pointer;
    background-color: rgba(128, 128, 128, 0.207);
  }
}

.text {
  display: grid;
  grid-template-columns: 1fr;
  gap: 1.2rem;
}

.textContainer1 {
  display: flex;
  align-items: center;
  gap: 0.7rem;
}

.profilePic {
  width: 60px;
  aspect-ratio: 1 / 1;
  border-radius: 50%;
  object-fit: cover;
}

.name {
  font-size: 1.1rem;
}

.createdAt {
  font-size: 0.9rem;
  color: rgb(128, 128, 128);
}

.title {
  font-size: 1.4rem;
}

.textContainer3 {
  overflow: hidden;
  box-shadow: inset 0 0 10px rgba(0, 0, 0, 0.435);
  padding: 0 0.5rem;
  max-height: 120px;
}

.textContainer3 > * {
  all: revert;
}

.textContainer4 {
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  gap: 1rem;
}

.textContainer4 > p {
  font-weight: 500;
  color: rgba(115, 113, 113, 0.761);
  background-color: rgba(128, 128, 128, 0.246);
  border-radius: 5px;
  padding: 0.3rem 0.6rem;
}

.textContainer5 {
  display: flex;
  align-items: center;
  gap: 1rem;
}

.textContainer5 strong {
  font-weight: 500;
}

.likeButton {
  background-color: black;
  mask-image: url("../../assets/images/heart.svg");
  mask-repeat: no-repeat;
  mask-size: contain;
  width: 25px;
  aspect-ratio: 1 / 1;
  cursor: pointer;
}

.likeButton.liked {
  background-color: crimson;
}

.loginMessage a {
  color: #000;
}

.imageContainer > img {
  width: 100%;
  height: 200px;
  object-fit: cover;
  border-radius: 5px;
}

@media (max-width: 650px) {
  .card {
    grid-template-columns: 1fr;
  }
}
