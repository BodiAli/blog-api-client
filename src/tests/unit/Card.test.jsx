import { screen, render } from "@testing-library/react";
import { createRoutesStub } from "react-router";
import Card from "../../components/Card/Card";
import { vi } from "vitest";

vi.stubGlobal("fetch", vi.fn);

const fakePost = {
  id: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
  userId: 1,
  title: "16",
  content: "<p>Content</p>",
  published: true,
  likes: 0,
  createdAt: "2025-05-01T10:44:36.648Z",
  updatedAt: "2025-05-01T10:44:36.648Z",
  imgUrl: null,
  User: {
    id: 1,
    firstName: "Bodi",
    lastName: "Ali",
    profileId: 1,
    Profile: {
      profileImgUrl: "img-url",
    },
  },
  Topics: [
    {
      id: 1,
      name: "Topic 1",
    },
  ],
};

const Stub = createRoutesStub([
  {
    path: "/",
    Component: () => <Card post={fakePost} />,
  },
]);

describe("Card component", () => {
  test("Should render post data", () => {
    const { container } = render(<Stub />);

    expect(container).toMatchSnapshot();
  });

  // test("Should ")
});
