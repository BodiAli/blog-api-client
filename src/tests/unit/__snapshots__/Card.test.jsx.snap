// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > Should render post data 1`] = `
<div>
  <div
    class="_card_1f2933"
  >
    <div
      class="_text_1f2933"
    >
      <div
        class="_textContainer1_1f2933"
      >
        <img
          alt="Bodi Ali's profile picture"
          class="_profilePic_1f2933"
          src="img-url"
        />
        <p
          class="_name_1f2933"
        >
          Bodi
           
          Ali
        </p>
        <p
          class="_createdAt_1f2933"
        >
          May 1, 2025
        </p>
      </div>
      <div
        class="_textContainer2_1f2933"
      >
        <p
          class="_title_1f2933"
        >
          16
        </p>
      </div>
      <div
        class="_textContainer3_1f2933"
      >
        <p>
          Content
        </p>
      </div>
      <div
        class="_textContainer4_1f2933"
      >
        <p>
          Topic 1
        </p>
      </div>
      <div
        class="_textContainer5_1f2933"
      >
        <p>
          <strong>
            Likes: 
          </strong>
          0
        </p>
      </div>
    </div>
    <div
      class="_imageContainer_1f2933"
    >
      <img
        alt="Post cover"
        src="/src/assets/images/no-image.svg"
      />
    </div>
  </div>
</div>
`;
